types/ndarray/ndarray-tests.ts

```diff
@@ -56,3 +56,11 @@ console.log(typeof firstVal === 'string' ? firstVal.length : firstVal.valueOf())
 function getFirstValue(arr: ndarray.NdArray): number {
     return arr.get(0);
 }
+
+function getFirstValueMaybeBigInt(arr: ndarray.NdArray<ndarray.Data<bigint>>): bigint | number {
```

Comment:
Then, hopefully, the type of `arr` could be written `ndarray.NdArray<bigint[]>` and the return type could just be `bigint`.

Reply:
> Then, hopefully, the type of arr could be written ndarray.NdArray<bigint[]> and the return type could just be bigint.

This already works with my implementation.

```typescript
function getFirstValueArray(arr: ndarray.NdArray<bigint[]>): bigint {
    return arr.get(0); // ok
}
```

The function you've highlighted is showing the explicit use of `ndarray.Data<bigint>` compared to `ndarray.Data<number>` to widen the inferred type to include `BigInt` as well (which is why it's called `getFirstValueMaybeBigInt`).

The previous type,

```typescript
type Value<D extends Data> = D extends Array<infer T> ? T : number;
```

hard-codes `TypedArray[number]` as number, so since `Data<T>` previously resolved to `T[] | TypedArray`, the current behavior of `Value<Data<T>>` is:

```typescript
Value<TypedArray | number[]> // number;
Value<Data<number>> // number
Value<Data<bigint>> // number | bigint
Value<Data<bigint | string>> // number | bigint | string
```

But this errors currently,

```typescript
Value<bigint[]> // bigint
Value<BigInt64Array> //  Type 'BigInt64Array' does not satisfy the constraint Data<any>
```

My implementation preserves the exact behavior as above,

```typescript
Value<TypedArray | number[]> // number;
Value<Data<number>> // number
Value<Data<bigint>> // number | bigint
Value<Data<bigint | string>> // number | bigint | string
```

but additionally,

```
Value<bigint[]> // bigint
Value<TypedArray | BigInt64Array | number[]> number | bigint
Value<BigInt64Array> //  bigint
```

Reply:
Maybe it's best to think of `T` as a switch:

- `Data<T (anything except bigint)>`:  T[] | TypedArray, so `Value<Data<T>>` == 'T | number'
- `Data<T (anthing including bigint)>`: T[] | TypedArray | BigInt64Array | BigUint64Array, so `Value<Data<T>>` == `T | number | bigint`.

Reply:
Right, I get you now, thanks for the explanations (and sorry for thinking you'd "kinda broken things"! :confounded:) I now get why `NdArray` defaulting to `NdArray<Data<number>>` led you to this elegant solution of making `bigint` support an opt-in through the `Data<bigint>` switch.

The thing is, I hadn't really meant for `Data<T>` to be used directly, so I hadn't considered the fact that `ndarray<Data<string>>([]).get(0)` returns `string | number`, which, to me, makes little sense.

Perhaps we should rethink the purpose of `Data<T>`... What if it included only the typed arrays that made sense for the given `T`?

```ts
Data<number> // number[] | TypedArray (this maintains backwards compatibility)
Data<bigint> // bigint[] | BigInt64Array | BigUint64Array
Data<number | bigint> // number[] | bigint[] | TypedArray | BigInt64Array | BigUint64Array
Data<string> // string[]
```

This way, `Value<D> = D[number]` (nice simplification, by the way :clap:) would make a lot more sense:

```
Value<Data<number>> // number
Value<Data<bigint>> // bigint
Value<Data<number | bigint>> // number | bigint
Value<Data<string>> // string
```

Reply:
Makes sense to me! Yes, the idea was to ideally have "bigint" be opt-in based on some type of switch. 

> Perhaps we should rethink the purpose of Data<T>... What if it included only the typed arrays that made sense for the given T?

Makes total sense to me! Implemented in the most recent commit. [playground](https://www.typescriptlang.org/play?#code/PQgEB4CcFMDNpgOwMbVAGwJYCMC8AiaAZ0WgA8AXAOm0wHNNEL9RgA+AKC4oE8AHNABEAhhWHgAKqFyhhiHm2mgOoUFPIVoiACZFQiAK4BbbAlAB+NQG0AuqAA+a-tG0BBSJGE9QALmWr1Si1dUFoGJgtrO0cAIXoASSYANgAWd09vWPoAVUYKVPSvXxU1UA1gvTC8h31jU0hIiVsaiWc3DyKsukT8tI7M0Di6XOS+jN9-KIBubmcnAXbx3BLHHoAOQp4V0B6ARiTN7Z6AZgAmQ9VHEYoN-u3rtYBhdGEjBYua6-2Pq7yzn9AADF0AB7UT-O6XIGg0QFfozVSqDi8ASgABqwnQBmg4EEZSCOj0IjEihkgishhMCBsMw46GgFFAYj8xPElPqbBm9MZ2GZoFZ4CqTE5dIZoCIFE8HhZonEEsgjDobCAA) if curious.